File: golang/backtrack.go
package main

import "log"

var N = 8

type Queen struct {
	Column int
	Row    int
}

func reject(candidate Queen, solution []Queen) bool {
	row, column := candidate.Row, candidate.Column
	for _, q := range solution {
		r, c := q.Row, q.Column
		if row == r ||
			column == c ||
			row+column == r+c ||
			row-column == r-c {
			return true
		}
	}
	return false
}

func children(parent Queen) chan Queen {
	column := parent.Column + 1
	c := make(chan Queen, 0)
	go func() {
		defer close(c)
		for r := 1; r < N+1; r++ {
			c <- Queen{column, r}
		}
	}()
	return c
}

func accept(solution []Queen) bool {
	return len(solution) == N
}

type RootStackEntry struct {
	Parent   Queen
	Children chan Queen
}

type RootStack []RootStackEntry

func rpush(s RootStack, e RootStackEntry) RootStack {
	return append(s, e)
}

func rpop(s RootStack) (RootStackEntry, RootStack) {
	if len(s) == 0 {
		return RootStackEntry{}, s
	}
	return s[len(s)-1], s[:len(s)-1]
}

type SolutionStack []Queen

func spush(s SolutionStack, e Queen) SolutionStack {
	return append(s, e)
}

func spop(s SolutionStack) (Queen, SolutionStack) {
	if len(s) == 0 {
		return Queen{}, s
	}
	return s[len(s)-1], s[:len(s)-1]
}

func backtrack(root Queen) {
	stack := RootStack{}
	solution := SolutionStack{}

	c := children(root)

	var candidate Queen
	var ok bool
	var rse RootStackEntry
	for {
		log.Println(solution)
		if candidate, ok = <-c; !ok {
			_, solution = spop(solution)
			if len(stack) == 0 {
				log.Println("NO ANSWER.")
				return
			}
			rse, stack = rpop(stack)
			root = rse.Parent
			c = rse.Children
			continue
		}
		if reject(candidate, solution) {
			continue
		}
		solution = spush(solution, candidate)
		if accept(solution) {
			log.Println("WINNER")
			log.Println(solution)
			return
		}
		stack = rpush(stack, RootStackEntry{root, c})
		root = candidate
		c = children(root)
	}
}

func init() {
	log.SetFlags(log.Lshortfile | log.LstdFlags)
}

func main() {
	backtrack(Queen{0, 0})
}

// 1, 0, 0, 0, 0, 0, 0, 0,
// 0, 0, 0, 0, 0, 0, 0, 0,
// 0, 0, 0, 0, 1, 0, 0, 0,
// 0, 0, 0, 0, 0, 0, 0, 0,
// 0, 1, 0, 0, 0, 0, 0, 0,
// 0, 0, 0, 1, 0, 0, 0, 0,
// 0, 0, 0, 0, 0, 1, 0, 0,
// 0, 0, 1, 0, 0, 0, 0, 0,
// [
// 1, 0, 0, 0,
// 0, 0, 0, 0,
// 0, 1, 0, 0,
// 0, 0, 0, 0,
// ]

File: python/backtrack.py
# def reject(solution, candidate):
#     column, row = candidate
#     return any(
#         row == r or
#         column == c or
#         row + column == r + c or 
#         row - column == r - c for r, c in solution)

# def accept(solution):
#     return globals(), locals(), len(solution) == N

# def children(parent):
#     column, _ = parent
#     for row in range(1, N + 1):
#         yield tuple([column + 1, row])

def add(solution, candidate):
    solution.append(candidate)

def remove(solution):
    solution.pop()

def output(solution):
    print(solution)

def __next(gen):
    try:
        return next(gen)
    except StopIteration:
        return None

# def backtrack(root, first, next, reject, accept, add=add, remove=remove, output=output, single_solution=False):
def backtrack(root):
    N = 4

    root_stack = list()
    solution = list()

    def reject(solution, candidate):
        column, row = candidate
        return any(
            row == r or
            column == c or
            row + column == r + c or 
            row - column == r - c for c, r in solution)

    def accept(solution):
        if len(solution) == N:
            print(solution)

    def children(parent):
        column, _ = parent
        for row in range(1, N + 1):
            yield tuple([column + 1, row])

    __children = children(root)

    while True:
        # print(root)
        candidate = __next(__children)
        if candidate is None:
            if len(root_stack) is 0:
                break
            remove(solution)
            root, __children = root_stack.pop()
            continue
        __reject = reject(solution, candidate)
        if __reject:
            continue
        add(solution, candidate)
        __accept = accept(solution)
        if __accept:
            remove(solution)
            continue
        root_stack.append([root, __children])
        root = candidate
        __children = children(root)

def go():
    backtrack((0, 1))

go()